### PR TITLE
Reduce use of LocalDefId <-> HirId maps

### PR DESCRIPTION
**Pull request by cjgillot**
This is an attempt to reduce the perf effect of https://github.com/rust-lang/rust/pull/89278.
r? @ghost

### CODE REVIEW

**Comment by cjgillot** [comment]
r? rust-lang/compiler

**Review by davidtwco — APPROVED** [review]
This looks reasonable to me, but probably warrants another opinion.

**Comment by davidtwco** [comment]
r? rust-lang/compiler

**Comment by nagisa** [comment]
Will not be able to take a look at this before 27th. Please reassign if this needs a quicker review.

File: compiler/rustc_middle/src/hir/map/mod.rs
Comment (nagisa) [inline compiler/rustc_middle/src/hir/map/mod.rs:356]: This function name suggests to me “get a def(id)” while what it really means is “get keyed by defid”. Would it be too onerous to make the name more accurate?

File: compiler/rustc_resolve/src/late/lifetimes.rs
Comment (nagisa) [inline compiler/rustc_resolve/src/late/lifetimes.rs:2030]: this is no longer printing the hir_id, is it? That said, `debug!(?def_id)` should be more future proof ^^

**Review by nagisa — APPROVED** [review]
Yeah, the idea seems sound. Couple nits inline, up to you if you think its worth to address them or not.

**Comment by nagisa** [comment]
I'm happy to r=me this. Please rebase and squash the history a little before merging though.

**Pull request merged**
